### ISSUE TITLE
Fluidsynth soundfonts not loading on Wii U(?)

### ISSUE THREAD

**Issue by CatmanFan** [EasyRPG/Player#3310]
#### Name of the game:
General, any game is applicable

#### Player platform:
Wii U

#### Attach files (as a .zip archive or link them)
 - A savegame next to the problem (if it is not at the beginning of the game)
*N/A* (not necessary for this issue)
 - The `easyrpg_log.txt` log file (you can find it in the game directory)
[easyrpg_log.txt](https://github.com/user-attachments/files/18089520/easyrpg_log.txt)

#### Describe the issue in detail and how to reproduce it:
1. Download and install the latest Wii U build (specifically, 2024-08-26):
https://easyrpg.org/player/downloads/#snapshot-other

2. Copy it to wiiu/apps/easyrpg-player/easyrpg-player.wuhb, then restart the console and open the EasyRPG icon from the Wii U Menu.

3. Go to *Settings > Audio*, enable `FluidLite (SF2)` from *MIDI drivers* and select any file or `<Autodetect>` from *Soundfont*.

4. Load a random game, after 2000 + 2003 RTP and some games have already been copied to the relevant EasyRPG directories.
The error message "*Debug: Fluidsynth: Could not load soundfont.*" is written to the log, and "Could not load soundfont" is always displayed at the bottom of the *Settings > Audio* submenu, regardless of whichever file has been loaded as the SF2 to use. FmMidi is thus enabled as the only audio driver available.

I've tested with the soundfont which was recommended by Ghabry from [this community post](https://community.easyrpg.org/t/configuring-midi-playback-on-web-player/875/2), and several downloaded from [this GitHub repo](https://github.com/Daniel-176/Usefull-Soundfonts), but none of them were able to work on the current Wii U build. At one point, `FluidLite (SF2)` was able to report itself as "working", but now it no longer is; in any case, FmMidi remained as the default audio driver.

No idea atm if this is related to some issue on the build's end or if FluidLite support has not been fully implemented yet.

**Comment by Ghabry** [comment]
Thanks for the report. Will try to get FluidSynth instead of FluidLite working for the WiiU. Lite is only used on platforms that lack RAM and the WiiU isn't one of them :).

(But the SF2 selection should still work with FluidLite, so that is a second bug :thinking: )

**Comment by Ghabry** [comment]
Good news: After a small patch because Threads are completely broken on WiiU fluidsynth compiles.

Cannot test it right now as don't have access to my WiiU but at least it compiles so when it works it will be in 0.8.1

```patch
diff -Naur fluidsynth-2.4.1-orig/CMakeLists.txt fluidsynth-2.4.1/CMakeLists.txt
--- fluidsynth-2.4.1-orig/CMakeLists.txt	2024-12-01 14:27:38.000000000 +0100
+++ fluidsynth-2.4.1/CMakeLists.txt	2024-12-21 21:34:58.813955060 +0100
@@ -355,14 +355,14 @@
 endif ( WIN32 )
 
 # Check for threads and math
-find_package ( Threads REQUIRED )
+#find_package ( Threads REQUIRED )
 
 find_library ( HAS_LIBM NAMES "m" )
 if ( HAS_LIBM )
   set ( MATH_LIBRARY "m" )
 endif ( HAS_LIBM )
 
-set ( LIBFLUID_LIBS ${MATH_LIBRARY} Threads::Threads )
+set ( LIBFLUID_LIBS ${MATH_LIBRARY} )
 
 # IBM OS/2
 unset ( DART_SUPPORT CACHE )
```

**Comment by Ghabry** [comment]
@CatmanFan can you try the latest continuous build? Should have FluidSynth enabled now instead of FluidLite.

Also cannot reproduce the problem with FluidLite. At least on Linux everything worked as expected.

**Comment by CatmanFan** [comment]
Nope, it's still the same issue: says it's "in use" when a certain SF2 is selected, "not working" when it's set to auto-detect, in either case it will always say "Could not load soundfont." and FmMidi will always load.

**Comment by Ghabry** [comment]
that patch is already applied in the Continuous Build. okay, weird. Have to test this myself on hardware I guess...

**Comment by Ghabry** [comment]
Lucky for us the WiiU supports remote debugging.

Loading of the soundfont fails because the chunksize is not matching the filesize.

![grafik](https://github.com/user-attachments/assets/fea0778e-8bfe-4492-91a2-37dc32027a8f)

Which indicates a lack of byte swapping because this is a big endian architecture.

Looking further at the code fluidsynth calls ``GINT32_FROM_LE`` which is stubbed in my shim... well... have to implement byte swapping I guess ;)

**Comment by Ghabry** [comment]
Updated patch adding the byte swapping:

```diff
diff '--color=auto' -Naur fluidsynth-2.4.2-orig/src/glib_shim.h fluidsynth-2.4.2/src/glib_shim.h
--- fluidsynth-2.4.2-orig/src/glib_shim.h	2025-01-15 20:37:59.253332471 +0100
+++ fluidsynth-2.4.2/src/glib_shim.h	2025-01-15 20:38:11.836473914 +0100
@@ -28,8 +28,13 @@
 #define G_BIG_ENDIAN 1
 #endif
 
+#ifdef WORDS_BIGENDIAN
+#define GINT32_FROM_LE(x) ((x >> 24) | ((x<<8) & 0x00FF0000) | ((x>>8) & 0x0000FF00) | (x << 24))
+#define GINT16_FROM_LE(x) ((x >> 8) | (x << 8))
+#else
 #define GINT32_FROM_LE(x) ((x))
 #define GINT16_FROM_LE(x) ((x))
+#endif
 
 #define G_LIKELY
 #define G_UNLIKELY
diff '--color=auto' -Naur fluidsynth-2.4.2-orig/src/sfloader/fluid_sffile.c fluidsynth-2.4.2/src/sfloader/fluid_sffile.c
--- fluidsynth-2.4.2-orig/src/sfloader/fluid_sffile.c	2025-01-15 20:37:59.249999190 +0100
+++ fluidsynth-2.4.2/src/sfloader/fluid_sffile.c	2025-01-15 20:39:15.682165703 +0100
@@ -2258,10 +2258,14 @@
     if(FLUID_IS_BIG_ENDIAN)
     {
         unsigned int i;
+        unsigned short data;
 
         for(i = 0; i < num_samples; i++)
         {
-            loaded_data[i] = FLUID_LE16TOH(loaded_data[i]);
+            data = loaded_data[i];
+            data = (data >> 8) | (data << 8);
+            //loaded_data[i] = FLUID_LE16TOH(loaded_data[i]);
+            loaded_data[i] = data;
         }
     }
 
```

**Comment by Ghabry** [comment]
Also upcoming while we discuss audio bugs here:

You are right, it is also broken for FluidLite which we still use on the Wii. Again endianess wrong.

OGG playback on WiiU is broken, again endianess (though here I have no idea why it is broken as vorbis does a runtime check for the endianess, added a fix to Player)

**Comment by Ghabry** [comment]
The latest Wii(U) continuous build has fixes for Fluidsynth.

Note that the Wii build has an inferior (lite) version so the soundfonts sound different. For WiiU it should match desktop quality.

The OGG fix is in a different (unmerged) PR.

Closing this. Please reopen when its still broken for you.
